ext/etc/etc.c

```diff
@@ -261,12 +394,59 @@ static VALUE
 etc_getpwnam(VALUE obj, VALUE nam)
 {
 #ifdef HAVE_GETPWENT
-    struct passwd *pwd;
-    const char *p = StringValueCStr(nam);
+# ifdef HAVE_GETPWNAM_R
+    char *login = RSTRING_PTR(nam);
+    char *bufnm;
+    long bufsizenm = GETPW_R_SIZE_INIT;  /* maybe -1 */
+
+    if (bufsizenm < 0)
+        bufsizenm = GETPW_R_SIZE_DEFAULT;
+
+    VALUE getpwnm_tmp = rb_str_tmp_new(bufsizenm);
+
+    bufnm = RSTRING_PTR(getpwnm_tmp);
+    bufsizenm = rb_str_capacity(getpwnm_tmp);
+    rb_str_set_len(getpwnm_tmp, bufsizenm);
+    struct getpwnam_r_args args = GETPWNAM_R_ARGS(login, bufnm, bufsizenm);
+
+    int enm;
+    errno = 0;
+    while ((enm = WITHOUT_GVL_INT(nogvl_getpwnam_r, &args)) != 0) {
+
+        if (enm == ENOENT || enm== ESRCH || enm == EBADF || enm == EPERM) {
+            /* not found; non-errors */
+            rb_str_resize(getpwnm_tmp, 0);
+            return Qnil;
+        }
 
-    pwd = getpwnam(p);
+        if (enm != ERANGE || bufsizenm >= GETPW_R_SIZE_LIMIT) {
```

Comment:
`bufsizenm` is invariant in this loop.

```suggestion
        if (enm != ERANGE || args.bufsize >= GETPW_R_SIZE_LIMIT) {
```